src/NUnitFramework/framework/Attributes/NonParallelizableAttribute.cs

```diff
@@ -0,0 +1,63 @@
+// ***********************************************************************
+// Copyright (c) 2017 Charlie Poole
+//
+// Permission is hereby granted, free of charge, to any person obtaining
+// a copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+// 
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+// 
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+// ***********************************************************************
+
+using System;
+using NUnit.Framework.Interfaces;
+using NUnit.Framework.Internal;
+using NUnit.Framework.Internal.Execution;
+
+namespace NUnit.Framework
+{
+    /// <summary>
+    /// ParallelizableAttribute is used to mark tests that may be run in parallel.
+    /// </summary>
+    [AttributeUsage( AttributeTargets.Assembly | AttributeTargets.Class | AttributeTargets.Method, AllowMultiple=false, Inherited=false )]
+    public sealed class NonParallelizableAttribute : PropertyAttribute
+    {
+        /// <summary>
+        /// Construct a NonParallelizableAttribute.
+        /// </summary>
+        public NonParallelizableAttribute()
+        {
+            Properties.Add(PropertyNames.ParallelScope, ParallelScope.None);
+        }
+
+        /// <summary>
+        /// Overridden to check for invalid combinations of settings
+        /// </summary>
+        /// <param name="test"></param>
+        public override void ApplyToTest(Test test)
+        {
+            base.ApplyToTest(test);
+
+            if (test.RunState != RunState.NotRunnable && test is TestAssembly)
+                MarkTestInvalid(test, "Assemblies may not be marked as NonParallel");
```

Comment:
Also, the question arises, what should we do if no attribute is present on the assembly? Is it better to start off on the non-parallel queue or the parallel queue? I guess it depends on what we are going to find at lower levels, which we can't know at the point of kicking things off.